### ISSUE TITLE
Accessible Authentication needs better techniques

### ISSUE THREAD

**Issue by mbgower** [w3c/wcag#1916]
[G128 Email link authentication](https://w3c.github.io/wcag/understanding/accessible-authentication.html) is the only existing technique for [Accessible Authentication](https://w3c.github.io/wcag/understanding/accessible-authentication.html). I don't believe anyone would recommend it as the optimal way of meeting this criterion, especially because it just 'kicks authentication down the road' by sending the password to an email account, where the user is going to have to authenticate to _that_ website, which in turn must meet the same criterion for accessible authentication.
Here are other candidates:
1. There are some obvious missing techniques that match each of the Examples: properly marking up inputs and not preventing paste, WebAuthn, etc.
1. A possible Advisory technique is: **Providing a mechanism for the user to reveal the masked password string** (something that allows users to confirm the populated field matches the password they expect)
1. A possible failure technique: **Failure of Success Criterion 3.3.7 due to restricting text input options**

It could focus on this kind of construction:
` <input type="text" onpaste="return false;" ondrop="return false;" autocomplete="off" />
myElement.addEventListener('paste', e => e.preventDefault());`

**Comment by alastc** [comment]
Yep, more techniques needed. Just need someone to draft one or more...

**Comment by mraccess77** [comment]
@mbgower with email auth you don't need to login again down the road - simply following the link in the email will log you in. An example that has this is Slack.  I believe Discord might have a similar one and one that you can log in from another device or through a QR code, etc.

**Comment by mbgower** [comment]
@mraccess77 
> with email auth you don't need to login again down the road - simply following the link in the email will log you in

My point would be you _do_ have to log into your email client to _get_ the link. So while the G128 technique may work with website X, it won't work with your email website (e.g., yahoo). You obviously need a different technique.

**Comment by patrickhlauke** [comment]
> you _do_ have to log into your email client

_if_ you're using a web-based email client, rather than a native one ...

**Comment by mraccess77** [comment]
@mbgower I was thinking native based email clients where the password is stored - but even with web based versions they often stay logged in for a period of time especially if you are logged into a browser like Chrome and use Gmail.

**Comment by mbgower** [comment]
> if you're using a web-based email client, rather than a native one ...

@patrickhlauke, right, and my original comment in a survey response captured that (is 'use a native email client' a technical solution?). But i suspect as a web standard we have to assume that our solution works in the context of a web environment.
 
 > I was thinking native based email clients where the password is stored 

@mraccess77 I think we'd have to specify that if that is the intention, and again i wonder if that is even 'allowed'. 

> they often stay logged in for a period of time

But you still have to log in once in a while AND arguably you would be more likely to forget your password in that scenario, and so the solution would be more critical

**Comment by patrickhlauke** [comment]
just a tangential observation:

>  properly marking up inputs and not preventing paste

are those two aspects intertwined/required to be together? I had the impression that those are two separate aspects. if they are indeed both necessary simultaneously, that wasn't clear in the wordings

**Comment by patrickhlauke** [comment]
> I don't believe anyone would recommend it as the optimal way of meeting this criterion

no, i think it's mostly that nobody's bothered to write any other techniques so far...

> assuming this gets ported to wcag2ict, you'd have to solve the same problem for your mobile device, operating system, etc.

of course in some of the techniques that are valid for web use case simply won't translate 1-to-1 to other more general ICT cases. also, self-evident situations like "if i can't remember my computer's login details, i can get an email, but since i can't log in that won't help me". don't think that that should invalidate the emailed magic link as a technique, particularly as it's far less burdensome for authors than, say, implementing a proper WebAuthn implementation or similar.

**Comment by alastc** [comment]
Yep, we do need people to step up and add techniques. I think raising an issue that says that is practically volunteering ;-) 

On the email topic, I think we should keep that aspect separate from the author responsibility. I.e. consider email a required technology, it is between the user and their email provider/client as to how they work.

**Comment by alastc** [comment]
@shawnthompson has volunteered to work on one, he'll be assigned when added to the repo.

**Comment by shawnthompson** [comment]
I can look into this

**Comment by JohnRochfordUMMS** [comment]
Shawn: See below a message from Steve Lee. The resource may be helpful for techniques.

Steve: Thank you!


John

John Rochford
University of Massachusetts Medical School
Eunice Kennedy Shriver Center
Director, INDEX Program
Faculty, Family Medicine & Community Health
DisabilityInfo.org<www.DisabilityInfo.org>
EasyText.AI<https://easytext.ai/>
LinkedIn<https://www.linkedin.com/in/john-rochford/>
About Me<https://johnrochford.com/?promo=email_sig&utm_source=product&utm_medium=email_sig&utm_campaign=edit_panel&utm_content=plaintext>
Schedule a meeting with me.<http://bit.ly/CallJR>

Confidentiality Notice:
This e-mail message, including any attachments, is for the sole use of the intended recipient(s) and may contain confidential, proprietary, and privileged information. Any unauthorized review, use, disclosure, or distribution is prohibited. If you are not the intended recipient, please contact the sender immediately and destroy or permanently delete all copies of the original message.

From: Steve Lee ***@***.******@***.***>
Sent: Wednesday, September 1, 2021 10:32 AM
To: Rochford, John ***@***.******@***.***>
Subject: How To Make Your Website's Authentication Process Accessible

Fyi https://www.boia.org/blog/is-your-sites-authentication-process-accessible<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fwww.boia.org%2Fblog%2Fis-your-sites-authentication-process-accessible&data=04%7C01%7Cjohn.rochford%40umassmed.edu%7C4d5330b235bf4cecdca908d96d55539a%7Cee9155fe2da34378a6c44405faf57b2e%7C0%7C0%7C637661036257025683%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C2000&sdata=MNdk1qyPMB7WEdAP6m8tCynvXx28fXhyXrFs8fyQ88Y%3D&reserved=0>

**Comment by JohnRochfordUMMS** [comment]
That another paat on help too.

**Comment by shawnthompson** [comment]
Thanks @JohnRochfordUMMS

**Comment by camusamta** [comment]
I just wanted to add a question to this thread about authenticator apps. The [current draft of 3.3.8](https://www.w3.org/WAI/WCAG22/Understanding/accessible-authentication) makes mention of multi-factor authentication (MFA) processes being a known part of modern authentication flows, and stipulates that for any authentication process:

> [that] relies on performing an action on a separate device, it should be possible to complete the action without the need to transcribe information. It may not be possible to know what device-based authentication methods are available to a user; offering a choice of methods can allow them to choose the path that most suits them.

I feel like, when testing any authentication process that involves either a user-elected authentication app, or an author-provided authenticator app, things get a bit murky. Most authenticator apps offer a time-based code which can be copied, but copying something from a separate device across to the website requiring that code can be impossible for some people, meaning they'll fall back on having to transcribe the code, which then fails this Success Criteria (SC).

However, there's also no way for me to know, as an auditor, if a user's particular authenticator app offers some alternative to that method, or allows the user to extend the code's lifetime. I also have no way to assess if an author-provided authentictor app is sufficient, because it could be argued its copy functionality allows for a method other than transcribing of the generated code, and that no extending of the time meets the 'essential' exception in [2.2.1 - Timing Adjustable](https://www.w3.org/WAI/WCAG21/Understanding/timing-adjustable.html). 

Because of this uncertainty, I feel like some general statement or decision about this context needs to make its way into the normative section of the SC. Without something more explicit, I don't feel like the scenario in which separate-device-MFA is mandatory is a clear one to resolve against the intentions of 3.3.8.

**Comment by alastc** [comment]
@shawnthompson have you made any progress on this one? Just reviewing issues and would like to resolve...

**Comment by dan-hw** [comment]
Would it be helpful to add failure techniques here too? It was good to see the "Enter the 3rd, 4th, and 6th character of your password" approach called out as a failure in the 'Intent' section of the latest working draft. 

If it's helpful to formalise this sort of approach as a Failure Technique, I'd be happy to submit a PR for it

**Comment by shawnthompson** [comment]
@mbgower, you commented in the survey:

> Don't we want something about not preventing pasting into the password field?

Do you think we should we add information about preventing the pasting into the description section of this technique and then add it into the test results?

**Comment by mbgower** [comment]
> Do you think we should we add information about preventing the pasting into the description section of this technique and then add it into the test results?

Yes, i think that's worth trying. To me, it's part of allowing email/password to be sufficient.

**Comment by dan-hw** [comment]
> Would it be helpful to add failure techniques here too? It was good to see the "Enter the 3rd, 4th, and 6th character of your password" approach called out as a failure in the 'Intent' section of the latest working draft.
> 
> If it's helpful to formalise this sort of approach as a Failure Technique, I'd be happy to submit a PR for it

I've submitted PR #2990 which adds a failure technique for this approach, with a few examples I've seen in the wild

**Comment by alastc** [comment]
New technique [agreed to be added yesterday](https://www.w3.org/2023/01/31-ag-minutes#item12).

**Comment by shawnthompson** [comment]
@alastc, I still need to make the recommended changes?

**Comment by mbgower** [comment]
@alastc I have reopened as there has been a failure technique created, the references this issue.
https://github.com/w3c/wcag/pull/2990
